engine/access/rpc/backend/connection_factory_test.go

```diff
@@ -407,6 +410,100 @@ func TestConnectionPoolStale(t *testing.T) {
 	assert.Equal(t, resp, expected)
 }
 
+// TestExecutionNodeClientClosedGracefully tests the scenario where the execution node client is closed gracefully.
+//
+// Test Steps:
+// - Generate a random number of requests and start goroutines to handle each request.
+// - Invalidate the execution API client.
+// - Wait for all goroutines to finish.
+// - Verify that the number of completed requests matches the number of sent responses.
+func TestExecutionNodeClientClosedGracefully(t *testing.T) {
+	// Add createExecNode function to recreate it each time for rapid test
+	createExecNode := func() (*executionNode, func()) {
+		en := new(executionNode)
+		en.start(t)
+		return en, func() {
+			en.stop(t)
+		}
+	}
+
+	// Add rapid test, to check graceful close on different number of requests
+	rapid.Check(t, func(t *rapid.T) {
+		en, closer := createExecNode()
+		defer closer()
+
+		// setup the handler mock to not respond within the timeout
+		req := &execution.PingRequest{}
+		resp := &execution.PingResponse{}
+		respSent := atomic.NewUint64(0)
+		en.handler.On("Ping", testifymock.Anything, req).Run(func(_ testifymock.Arguments) {
+			respSent.Inc()
+		}).Return(resp, nil)
+
+		// create the factory
+		connectionFactory := new(ConnectionFactoryImpl)
+		// set the execution grpc port
+		connectionFactory.ExecutionGRPCPort = en.port
+		// set the execution grpc client timeout
+		connectionFactory.ExecutionNodeGRPCTimeout = time.Second
+		// set the connection pool cache size
+		cacheSize := 1
+		cache, _ := lru.NewWithEvict(cacheSize, func(_, evictedValue interface{}) {
+			evictedValue.(*CachedClient).Close()
+		})
+		connectionFactory.ConnectionsCache = cache
+		connectionFactory.CacheSize = uint(cacheSize)
+		// set metrics reporting
+		connectionFactory.AccessMetrics = metrics.NewNoopCollector()
+
+		clientAddress := en.listener.Addr().String()
+		// create the execution API client
+		client, _, err := connectionFactory.GetExecutionAPIClient(clientAddress)
+		assert.NoError(t, err)
+
+		result, _ := cache.Get(clientAddress)
+		clientConn := result.(*CachedClient).ClientConn
+		clientConn.GetState()
+
+		ctx := context.Background()
+
+		// Generate random number of requests
+		nofRequests := rapid.IntRange(10, 100).Draw(t, "nofRequests").(int)
+		reqCompleted := atomic.NewUint64(0)
+
+		var waitGroup sync.WaitGroup
+
+		for i := 0; i < nofRequests; i++ {
+			waitGroup.Add(1)
+
+			// call Ping request from different goroutines
+			go func() {
+				defer waitGroup.Done()
+				_, err := client.Ping(ctx, req)
+
+				if err == nil {
+					reqCompleted.Inc()
+				} else {
+					if st, ok := status.FromError(err); ok {
+						if st.Code() == codes.Unavailable {
+							return
+						}
+					}
+					fmt.Println("!!! Failed: ", err)
+					t.Fail()
```

Comment:
You can simplify this check a bit.
```suggestion
					require.Equalf(t, codes.Unavailable, status.Code(err), "unexpected error: %v", err)
```